pkg/services/query/query.go

```diff
@@ -151,24 +161,32 @@ func (s *ServiceImpl) executeConcurrentQueries(ctx context.Context, user *user.S
 	}
 	close(rchan)
 	resp := backend.NewQueryDataResponse()
+	reqCtx := contexthandler.FromContext(ctx)
 	for result := range rchan {
-		for refId, dataResponse := range result {
+		for refId, dataResponse := range result.responses {
 			resp.Responses[refId] = dataResponse
 		}
+		if reqCtx != nil {
+			for k, v := range result.header {
+				for _, s := range v {
+					reqCtx.Resp.Header().Add(k, s)
```

Comment:
Think this should use Set instead of Add or check if not exists before adding it. Or maybe check if both key and value not exist and add it. Not sure, but might introduce some weird behavior?

Reply:
Added a check for duplicates and a warning here [258399f](https://github.com/grafana/grafana/pull/70100/commits/258399fd6126db95dbf4c0f133ad6a2675ad92eb)